### ISSUE TITLE
Gracy gives up after first retry on 504 error when a NonOkResponse is raised

### ISSUE THREAD

**Issue by mborus** [guilatrova/gracy#9]
Hi,

I consistenly have a problem with one API where Gracy does only retry once and then quits with a **NonOkResponse** error.

What happens is that I query an API very slowly, once every 2 seconds, but sometimes the API goes down and returns a https status code 504.
What should happen is 5 retries after a long wait between each. What happens is that the first retry immediately crashes.

Note: I remove the non-public parts from the URL with "REMOVED"

    NonOkResponse: https://<REMOVED>/2023-03-02%2018%3A00%3A00/24 raised 504, but it was expecting any successful status code
      File "radio.py", line 183, in main
        pl = await playlistapi.get_playlist(<removed>, playlist_dt=playlist_dt_hour)
      File "radio.py", line 149, in get_playlist
        return await self.get(PlaylistApiEndpoint.GET_PLAYLIST, params)
      File "gracy/_core.py", line 312, in get
        return await self._request("GET", endpoint, endpoint_args, *args, **kwargs)
      File "gracy/_core.py", line 303, in _request
        return await graceful_request
      File "gracy/_core.py", line 227, in _gracify
        raise validation_exc
      File "gracy/_core.py", line 201, in _gracify
        validator.check(result)
      File "gracy/_validators.py", line 15, in check
        raise NonOkResponse(str(response.url), response)

The time flow is like this.

1. Gracy does the request, which times out after almost a minute

 "took 0:00:59.626751"

2. Gracy schedules a retry 30 seconds later

        GracefulRetry: https://<REMOVED>/2023-03-02%2018%3A00%3A00/24 will wait 30s before next attempt (1 out of 5)
    
    {asctime: 2023-03-03 00:47:26,091, CUR_ATTEMPT: 1, ENDPOINT: <REMOVED>/2023-03-02%2018%3A00%3A00/24, MAX_ATTEMPT: 5, METHOD: GET, RETRY_DELAY: 30, UENDPOINT:<REMOVED>/{datestr}%20{hour}%3A00%3A00/24, URL: https://<REMOVED>/2023-03-02%2018%3A00%3A00/24
    
    , UURL: https://<REMOVED>/{datestr}%20{hour}%3A00%3A00/24
    
    }

3. Gracy retries exactly 30 seconds later

    GracefulRetry: https:/<REMOVED>/2023-03-02%2018%3A00%3A00/24 attempt (1 out of 5)
    
    {asctime: 2023-03-03 00:47:56,315, CUR_ATTEMPT: 1, ENDPOINT: <REMOVED>2023-03-02%2018%3A00%3A00/24, MAX_ATTEMPT: 5, METHOD: GET, RETRY_DELAY: 30, UENDPOINT:<REMOVED>/{datestr}%20{hour}%3A00%3A00/24, URL: https://<REMOVED>/2023-03-02%2018%3A00%3A00/24
    
    , UURL: https://<REMOVED>/{datestr}%20{hour}%3A00%3A00/24
    
    }    

4. Within the same second, the error occurs
    NonOkResponse: https://<removed>/2023-03-02%2018%3A00%3A00/24 raised 504, but it was expecting any successful status code

The API is defined this way

```py
    # 1. Define your endpoints
    class PlaylistApiEndpoint(BaseEndpoint):
        GET_PLAYLIST = "<removed>/{datestr}%20{hour}%3A00%3A00/24"  

    # 2. Define your Graceful API
    class PlaylistAPI(Gracy[str]):
        class Config:  
            BASE_URL = "https://<REMOVED>"
            
            SETTINGS = GracyConfig(
                log_request=LogEvent(LogLevel.DEBUG),
                log_response=LogEvent(LogLevel.INFO, "{URL} took {ELAPSED}"),
                parser={
                    "default": lambda r: r.json()
                },
                throttling=GracefulThrottle(
                    rules=ThrottleRule(url_pattern=r".*",
                                       max_requests=1,
                                       per_time=datetime.timedelta(seconds=2)),
                ),
                retry=GracefulRetry(
                    delay=30,
                    max_attempts=5,
                    delay_modifier=1.5,
                    retry_on=None,
                    log_before=LogEvent(LogLevel.INFO),
                    log_after=LogEvent(LogLevel.WARNING),
                    log_exhausted=LogEvent(LogLevel.CRITICAL),
                    behavior="break",
                )
            )

        async def get_playlist(self,
                               # <REMOVED>
                               playlist_dt: datetime.datetime,
                               ) -> Awaitable[dict]:
            params = {
                "<REMOVED>":"",
                "datestr": f"{playlist_dt:%Y-%m-%d}",
                "hour": f"{playlist_dt:%H}"}

            return await self.get(PlaylistApiEndpoint.GET_PLAYLIST, params)
```

Thoughts: Since the first retry always crashes immediately, I don't trust the 504 status_code error message fully: Because the error comes so quickly - it's totally possible that the API returned a correct result with a 200 status_code on the retry. I also don't know yet if I had any successful retries with this API.

I'm logging/following this problem in Sentry. It occurs a few time each day and I will update this issue if I get more ideas or info.

Also I saw your answer on the issue https://github.com/guilatrova/gracy/issues/7 and will implement that, still I was expecting a 504 (Gateway Timeout) to trigger a retry instead of attempting a parse.

**Comment by guilatrova** [comment]
hmmm good point, I'll test it a bit more to ensure I cover this scenario

**Comment by guilatrova** [comment]
Your theory was right.

Gracy retried the request while failing but completely ignored the last successful result.

Fixed in version: [v1.11.2](https://github.com/guilatrova/gracy/blob/main/CHANGELOG.md#v1112-2023-03-03)

**Comment by mborus** [comment]
Just some quick feedback - this worked well for 10 retry attempts today on the API that would before give me problems. Thanks.
